test/widgets/compose_box_test.dart

```diff
@@ -0,0 +1,106 @@
+import 'package:checks/checks.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/widgets/compose_box.dart';
+
+void main() {
+  group('ComposeContentController', () {
+    group('insertPadded', () {
+      // Like `parseMarkedText` in test/model/autocomplete_test.dart,
+      //   but a bit different -- could maybe deduplicate some.
+      parseMarkedText(String markedText) {
+        final textBuffer = StringBuffer();
+        int? insertionPoint;
+        int i = 0;
+        for (final char in markedText.codeUnits) {
+          if (char == 94 /* ^ */) {
+            if (insertionPoint != null) {
+              throw Exception('Test error: too many ^ in input');
+            }
+            insertionPoint = i;
+            continue;
+          }
+          textBuffer.writeCharCode(char);
+          i++;
+        }
+        if (insertionPoint == null) {
+          throw Exception('Test error: expected ^ in input');
+        }
+        return TextEditingValue(text: textBuffer.toString(), selection: TextSelection.collapsed(offset: insertionPoint));
+      }
+
+      /// Test the given `insertPadded` call, in a convenient format.
+      ///
+      /// In valueBefore, represent the insertion point as "^".
+      /// In expectedValue, represent the collapsed selection as "^".
+      checkInsertPadded(String description, String valueBefore, String textToInsert, String expectedValue) {
+        test(description, () {
+          final controller = ComposeContentController();
+          controller.value = parseMarkedText(valueBefore);
+          controller.insertPadded(textToInsert);
+          check(controller.value).equals(parseMarkedText(expectedValue));
+        });
+      }
+
+      // TODO(?) exercise the part of insertPadded that chooses the insertion
+      //   point based on [TextEditingValue.selection], which may be collapsed,
+      //   expanded, or null (what they call !TextSelection.isValid).
+
+      checkInsertPadded('empty; insert one line',
+        '^', 'a\n',    'a\n\n^');
+      checkInsertPadded('empty; insert two lines',
+        '^', 'a\nb\n', 'a\nb\n\n^');
+
+      group('insert at end', () {
+        checkInsertPadded('one empty line; insert one line',
```

Comment:
Let's have a convention that when the helper creates its own test cases (calls `test`), its name is like `testFoo`, and `checkFoo` is for helpers one calls inside an existing test case. That will help with correctly keeping all the test computations inside test cases.